File: python/src/smooai_config/cloud_region.py
"""Cloud provider and region detection from environment variables."""

import os
from dataclasses import dataclass


@dataclass
class CloudRegionResult:
    """Result of cloud provider/region detection."""

    provider: str
    region: str


def get_cloud_region(env: dict[str, str] | None = None) -> CloudRegionResult:
    """Detect cloud provider and region from environment variables.

    Detection order:
    1. SMOOAI_CONFIG_CLOUD_REGION / SMOOAI_CONFIG_CLOUD_PROVIDER (custom override)
    2. AWS_REGION / AWS_DEFAULT_REGION
    3. AZURE_REGION / AZURE_LOCATION
    4. GOOGLE_CLOUD_REGION / CLOUDSDK_COMPUTE_REGION
    5. Default: unknown/unknown

    Args:
        env: Environment variable dict. If None, uses os.environ.
    """
    if env is None:
        env = dict(os.environ)

    # 1. Custom override
    if env.get("SMOOAI_CONFIG_CLOUD_REGION") or env.get("SMOOAI_CONFIG_CLOUD_PROVIDER"):
        return CloudRegionResult(
            provider=env.get("SMOOAI_CONFIG_CLOUD_PROVIDER", "unknown"),
            region=env.get("SMOOAI_CONFIG_CLOUD_REGION", "unknown"),
        )

    # 2. AWS
    aws_region = env.get("AWS_REGION") or env.get("AWS_DEFAULT_REGION")
    if aws_region:
        return CloudRegionResult(provider="aws", region=aws_region)

    # 3. Azure
    azure_region = env.get("AZURE_REGION") or env.get("AZURE_LOCATION")
    if azure_region:
        return CloudRegionResult(provider="azure", region=azure_region)

    # 4. GCP
    gcp_region = env.get("GOOGLE_CLOUD_REGION") or env.get("CLOUDSDK_COMPUTE_REGION")
    if gcp_region:
        return CloudRegionResult(provider="gcp", region=gcp_region)

    # 5. Default
    return CloudRegionResult(provider="unknown", region="unknown")

File: dotnet/src/SmooAI.Config/Config/CloudRegion.cs
namespace SmooAI.Config.Config;

/// <summary>The detected cloud provider and region.</summary>
/// <param name="Provider">One of <c>aws</c>, <c>azure</c>, <c>gcp</c>, a custom override, or <c>unknown</c>.</param>
/// <param name="Region">The provider's region identifier, or <c>unknown</c>.</param>
public sealed record CloudRegionResult(string Provider, string Region);

/// <summary>
/// Detects the cloud provider and region from environment variables.
/// </summary>
/// <remarks>
/// Port of <c>src/config/getCloudRegion.ts</c> and <c>go/config/cloud_region.go</c>.
/// </remarks>
public static class CloudRegion
{
    /// <summary>Value reported when nothing identifies the environment.</summary>
    public const string Unknown = "unknown";

    /// <summary>Detect from the process environment.</summary>
    /// <returns>The detected provider and region.</returns>
    public static CloudRegionResult Get() => GetFromEnv(ReadProcessEnvironment());

    /// <summary>
    /// Detect from an explicit environment map.
    /// </summary>
    /// <param name="env">The environment to inspect.</param>
    /// <returns>The detected provider and region.</returns>
    /// <remarks>
    /// Detection order, first hit wins:
    /// <list type="number">
    /// <item><description><c>SMOOAI_CONFIG_CLOUD_REGION</c> / <c>SMOOAI_CONFIG_CLOUD_PROVIDER</c> (custom override)</description></item>
    /// <item><description><c>AWS_REGION</c> / <c>AWS_DEFAULT_REGION</c></description></item>
    /// <item><description><c>AZURE_REGION</c> / <c>AZURE_LOCATION</c></description></item>
    /// <item><description><c>GOOGLE_CLOUD_REGION</c> / <c>CLOUDSDK_COMPUTE_REGION</c></description></item>
    /// <item><description>otherwise <c>unknown</c>/<c>unknown</c></description></item>
    /// </list>
    /// </remarks>
    public static CloudRegionResult GetFromEnv(IReadOnlyDictionary<string, string?> env)
    {
        ArgumentNullException.ThrowIfNull(env);

        // 1. Custom override — either half is enough to select this branch, so
        //    a provider can be named without pinning a region and vice versa.
        var overrideRegion = Read(env, "SMOOAI_CONFIG_CLOUD_REGION");
        var overrideProvider = Read(env, "SMOOAI_CONFIG_CLOUD_PROVIDER");
        if (overrideRegion is not null || overrideProvider is not null)
        {
            return new CloudRegionResult(overrideProvider ?? Unknown, overrideRegion ?? Unknown);
        }

        // 2. AWS
        if (Coalesce(env, "AWS_REGION", "AWS_DEFAULT_REGION") is { } aws) return new CloudRegionResult("aws", aws);

        // 3. Azure
        if (Coalesce(env, "AZURE_REGION", "AZURE_LOCATION") is { } azure) return new CloudRegionResult("azure", azure);

        // 4. GCP
        if (Coalesce(env, "GOOGLE_CLOUD_REGION", "CLOUDSDK_COMPUTE_REGION") is { } gcp) return new CloudRegionResult("gcp", gcp);

        return new CloudRegionResult(Unknown, Unknown);
    }

    private static string? Read(IReadOnlyDictionary<string, string?> env, string name)
        => env.TryGetValue(name, out var value) && !string.IsNullOrEmpty(value) ? value : null;

    private static string? Coalesce(IReadOnlyDictionary<string, string?> env, params string[] names)
    {
        foreach (var name in names)
        {
            if (Read(env, name) is { } value) return value;
        }
        return null;
    }

    internal static Dictionary<string, string?> ReadProcessEnvironment()
    {
        var result = new Dictionary<string, string?>(StringComparer.Ordinal);
        foreach (System.Collections.DictionaryEntry entry in Environment.GetEnvironmentVariables())
        {
            if (entry.Key is string key) result[key] = entry.Value as string;
        }
        return result;
    }
}
